### DISCUSSION TITLE
Batching RabbitListener not working List<Message<Object>> messages

### DISCUSSION THREAD

**Discussion by adityamparikh** [spring-projects/spring-amqp#2559] [Q&A] [answered]
Had a Spring Amqp Question. When enabling batching

```
    @Bean
    public RabbitListenerContainerFactory<?> batchContainerFactory(ConnectionFactory connectionFactory) {
        SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
        factory.setConnectionFactory(connectionFactory);
        factory.setAcknowledgeMode(AcknowledgeMode.MANUAL); // Enable manual acknowledgement
        factory.setBatchListener(true);
        factory.setBatchSize(1024); // Set the number of messages to be received in a batch
        factory.setDeBatchingEnabled(false);
        return factory;
    }
 @RabbitListener(queues = "#{queue.getName()}", containerFactory = "batchContainerFactory")
    public void handleBatch(List<Message<Object>> messages, Channel channel) throws IOException {}.
```

In the RabbitListener I directly get the payload of the message
and get class java.lang.String cannot be cast to class org.springframework.messaging.Message (java.lang.String is in module java.base of loader 'bootstrap'; org.springframework.messaging.Message is in unnamed module of loader 'app') Tried with Spring boot 2.7 as well as 3.1.5

**Comment by garyrussell** [comment, marked as answer]
You also need to set `consumerBatchEnabled` to enable batching.

```java
factory.setConsumerBatchEnabled(true);
```

```java
	/**
	 * Set to true to present a list of messages based on the {@link #setBatchSize(Integer)},
	 * if the listener supports it. Starting with version 3.0, setting this to true will
	 * also {@link #setBatchListener(boolean)} to true.
	 * @param consumerBatchEnabled true to create message batches in the container.
	 * @since 2.2
	 * @see #setBatchSize(Integer)
	 * @see #setBatchListener(boolean)
	 */
	public void setConsumerBatchEnabled(boolean consumerBatchEnabled) {
```

**Reply by adityamparikh** [reply]
Thanks @garyrussell This got me past the cast error. However, the Batch Listener seems to be doing escaping that the regular listener is not.
expected: <{"ticker":"ABC"}> but was: <"{\"ticker\":\"ABC\"}">
For a simple test object record Trade(String ticker){};

**Reply by garyrussell** [reply]
That doesn't make any sense to me; what happens if you consume `List<Object>` ?

Can you show a screenshot of a message from the RabbitMQ management console?

**Reply by adityamparikh** [reply]
byte[] worked. I had to override the RabbitTemplate and RabbitMessagingTemplate for my test as I was overriding the MessageConverter.
